Add solution for 4-module 2-task

akoltun opened the pull request.

koltual1 committed:
Add solution for 4-module 2-task

jsru commented:
Проверьте, пожалуйста, ваше решение, не все тесты прошли (PR не будет принят до тех пор, пока **все** добавленные задачи не будут решены).

```diff
Файловый сервер - запись файла
+ возвращается ошибка 409 при создании файла, который есть
+ при попытке создания слишком большого файла - ошибка 413
+ успешное создание файла
+ файл не должен оставаться на диске при обрыве соединения
+ если путь вложенный - возвращается ошибка 400
- если тело запроса пустое файл не перезаписывается
```

jsru commented:
Проверьте, пожалуйста, ваше решение, не все тесты прошли (PR не будет принят до тех пор, пока **все** добавленные задачи не будут решены).

```diff
Файловый сервер - запись файла
+ возвращается ошибка 409 при создании файла, который есть
+ при попытке создания слишком большого файла - ошибка 413
+ успешное создание файла
+ файл не должен оставаться на диске при обрыве соединения
+ если путь вложенный - возвращается ошибка 400
- если тело запроса пустое файл не перезаписывается
```

koltual1 committed:
Add solution for 4-module 3-task

jsru commented:
Проверьте, пожалуйста, ваше решение, не все тесты прошли (PR не будет принят до тех пор, пока **все** добавленные задачи не будут решены).

```diff
Файловый сервер - запись файла
+ возвращается ошибка 409 при создании файла, который есть
+ при попытке создания слишком большого файла - ошибка 413
+ успешное создание файла
+ файл не должен оставаться на диске при обрыве соединения
+ если путь вложенный - возвращается ошибка 400
- если тело запроса пустое файл не перезаписывается
```
```diff
Файловый сервер - удаление файла
+ файл должен удаляться
+ если файла нет - ошибка 404
+ если путь вложенный - возвращается ошибка 400
```

akoltun committed:
Fix solution for 4-module 2-task so it works with Node versions less than 12.13

When error happened,
- old Nodes (versions up to 12.12) emit finish event and then error event;
- new Nodes (starting from 12.13) emit only error event (no finish anymore).

jsru commented:
Проверьте, пожалуйста, ваше решение, не все тесты прошли (PR не будет принят до тех пор, пока **все** добавленные задачи не будут решены).

```diff
Файловый сервер - запись файла
+ возвращается ошибка 409 при создании файла, который есть
+ если тело запроса пустое файл не перезаписывается
+ при попытке создания слишком большого файла - ошибка 413
+ файл не должен оставаться на диске при обрыве соединения
+ если путь вложенный - возвращается ошибка 400
- успешное создание файла
- "after all" hook in "тесты на ��айловый сервер"
```
```diff
Файловый сервер - удаление файла
+ файл должен удаляться
+ если файла нет - ошибка 404
+ если путь вложенный - возвращается ошибка 400
```

akoltun committed:
Fix solution for 4-module 2-task specifically for Node 12.4 because it's Writable Stream doesn't have writableFinished property

jsru commented:
Поздравляю, все задачи решены абсолютно верно
```diff
Файловый сервер - запись файла
+ возвращается ошибка 409 при создании файла, который есть
+ если тело запроса пустое файл не перезаписывается
+ при попытке создания слишком большого файла - ошибка 413
+ успешное создание файла
+ файл не должен оставаться на диске при обрыве соединения
+ если путь вложенный - возвращается ошибка 400
```
```diff
Файловый сервер - удаление файла
+ файл должен удаляться
+ если файла нет - ошибка 404
+ если путь вложенный - возвращается ошибка 400
```
